### ISSUE TITLE
Extremely Slow Network (Regression) Training on Spark

### ISSUE THREAD

**Issue by ansarisam** [deeplearning4j/deeplearning4j#6437]
Spark Cluster Size: 3-node with each node having 32GB RAM and 8-core CPU
Total Spark Core: 20
Data Size: 45 million rows with size 9GB
Data is read and split pretty fast (within 1 minute)
After that the process moves very slow. Spark Stage DAG Visualizer shows the following:
![image](https://user-images.githubusercontent.com/26036956/45532707-02e70380-b7c3-11e8-9ac6-b097ce088d0b.png)

This is happening even before starting to train the network.  the slowness is observed at:
MapPartitionsRDD [44]saveAsTextFile at BaseTrainingMaster.java:184 

Looks like the writing TextFile to HDFS is extremely slow.
The following image should give an idea that it was able to process 64.4MB in 42 minutes (in line 1).
![image](https://user-images.githubusercontent.com/26036956/45532855-86a0f000-b7c3-11e8-9abe-70c368c32b4b.png)

Is there any way to avoid this expensive write? Can we cache the RDD to avoid this write to HDFS?

**Comment by ansarisam** [comment]
Yes, I am using 1.0.0-beta (not beta2). Here is the code:

`JavaRDD<DataSet> scaledDataSet = (new DataHandler()).convertToDataSet(scaledTrainingData2, conf, outputColumn, scaledLabelColumn);

// Split the dataset into training and test sets (70:30)
            JavaRDD<DataSet>[] splits = scaledDataSet.randomSplit(new double[]{0.7, 0.3});
            JavaRDD<DataSet> trainingData = splits[0];
            JavaRDD<DataSet> testData = splits[1];


// Define the Network
            // Number of features decides the nodes in the input layer.
            String[] featureFields = processfields.split(",");
            //Create network configuration and conduct network training
            NeuralNetConfiguration.ListBuilder builder = new NeuralNetConfiguration.Builder()
                    .seed(21)
                    .weightInit(WeightInit.XAVIER)
                    .updater(Nesterovs.builder().learningRate(lr).build())
                    .list();
            // specify layers for the neural network:
            // input layer = number of features, intermediate layers having nodes of size equal to numNodeHiddenLayer

            for (int i = 0; i < numLayer; i++) {
                Layer layer = null;
                if (i == 0) { // input layer
                    if (modelType.equalsIgnoreCase("class")) {
                        layer = new DenseLayer.Builder().nIn(featureFields.length).nOut(numNodeHiddenLayer).build();
                    } else {
                        layer = new DenseLayer.Builder().nIn(featureFields.length).nOut(numNodeHiddenLayer)
                                .activation(Activation.IDENTITY).build();
                    }

                    builder.layer(layer);
                } else if (i == numLayer - 1) { // output layer
                    if (modelType.equalsIgnoreCase("class")) {
                        layer = new OutputLayer.Builder(LossFunctions.LossFunction.NEGATIVELOGLIKELIHOOD).activation(Activation.SIGMOID).nIn(numNodeHiddenLayer).nOut(maxNumCategory).build();
                    } else {
                        layer = new OutputLayer.Builder(LossFunctions.LossFunction.MSE)
                                .activation(Activation.IDENTITY)
                                .nIn(numNodeHiddenLayer).nOut(maxNumCategory).build();
                    }
                    builder.layer(layer);
                } else { // hidden layers
                    if (modelType.equalsIgnoreCase("class")) {
                        layer = new DenseLayer.Builder().nIn(numNodeHiddenLayer).nOut(numNodeHiddenLayer).build();
                    } else {
                        layer = new DenseLayer.Builder().nIn(numNodeHiddenLayer).nOut(numNodeHiddenLayer)
                                .activation(Activation.IDENTITY).build();
                    }

                    builder.layer(layer);
                }
            }
            builder.pretrain(false).backprop(true);

            MultiLayerConfiguration netConf = builder.build();

            //Configuration for Spark training: see http://deeplearning4j.org/spark for explanation of these configuration options
            TrainingMaster tm = new ParameterAveragingTrainingMaster.Builder(batchSizePerWorker) 
                    .averagingFrequency(averagingFrequency)
                    .workerPrefetchNumBatches(workerPrefetchNumBatches) 
                    .batchSizePerWorker(batchSizePerWorker)
                    .saveUpdater(true)
                    .build();

            //Create the Spark network
            SparkDl4jMultiLayer sparkNet = new SparkDl4jMultiLayer(sc, netConf, tm);
// Fit the ANN model using training set


            for (int i = 0; i < iteration; i++) {
                sparkNet.fit(trainingData);
                }`

**Comment by AlexDBlack** [comment]
OK, so 2 things at this point.
First: what you are seeing is an export step before training. That 'saveAsTextFile' is saving the RDD<DataSet> to disk, and returning their paths as an RDD<String>.
Can you skip it? yes, by setting direct training mode: ```.rddTrainingApproach(RDDTrainingApproach.Direct)```
Should you? No, there's good performance reasons for exporting first.

Second: I suspect the slowness would be coming from your code:
```
JavaRDD scaledDataSet = (new DataHandler()).convertToDataSet(scaledTrainingData2, conf, outputColumn, scaledLabelColumn);
```
Of course, I can't say anything about this without actually seeing what it does.

What I suggest you do at this point:
1. Upgrade to 1.0.0-beta2. There were some spark improvements, and I'm hesitant to debug an old version
2. Preprocess your data once only, and use the saved data in later fit operations

For preprocessing: use this method.
https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j/deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/data/BatchAndExportDataSetsFunction.java

Then for fitting, call fit(String path):
https://github.com/deeplearning4j/deeplearning4j/blob/0294e884659c05dc8cdbb9aef5cce0ff10e420e2/deeplearning4j/deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/impl/multilayer/SparkDl4jMultiLayer.java#L257-L265

Finally, you might consider using SharedTrainingMaster. Your problem isn't caused or related to ParameterAveragingTrainingMaster, but SharedTrainingMaster will give you better training performance once you get to that point.
https://github.com/deeplearning4j/dl4j-examples/blob/master/dl4j-spark-examples/dl4j-spark/src/main/java/org/deeplearning4j/patent/TrainPatentClassifier.java#L191-L197

**Comment by ansarisam** [comment]
Thanks Alex. I will upgrade to -beta2 and follow your suggestions.
 `JavaRDD scaledDataSet = (new DataHandler()).convertToDataSet(scaledTrainingData2, conf, outputColumn, scaledLabelColumn);
`
FYI.. The above step is not an issue. I had a few log statements that I removed to avoid the clutter in the code snippet that I provided. I can see the above step is passing by very fast. Also the next line is about splitting the data into training and test sets -- thats also moving pretty fast.

**Comment by AlexDBlack** [comment]
>  I can see the above step is passing by very fast. Also the next line is about splitting the data into training and test sets -- thats also moving pretty fast.

Unfortunately Spark doesn't work like that.
Spark performs lazy evaluation - so x.map(...).map(...).map(...) won't actually do anything. It'll appear to return almost immediately.
Yet when you finally do something that *does* require calculation (saving, exporting, forEach, etc) it'll start execution of all steps. Hence if saveAsTextFile is slow, it can mean that *any* of the earlier steps are slow.

**Comment by ansarisam** [comment]
I had a a few log statements that prints count using .count(), have .show() and .collect statements which worked fast enough.

**Comment by AlexDBlack** [comment]
Another strategy here would be to run it locally on a single node (i.e., spark local) and attach a profiler such as YourKit profiler or VisualVM. (You can attach a profiler to cluster jobs, but it's difficult - probably more than it works).
Profiler screenshots (or better yet - an actual dump) would  help us narrow this down.

**Comment by Charele** [comment]
I see the code,there is a procedure called BatchAndExportDataSetsFunction,this change DataSet to storage format. As Alex said, RDD is lazy evaluation,  your pre-process(like "convertToDataSet", or others) will begin execute in BatchAndExportDataSetsFunction.

So I think you can also test this step:
...
JavaRDD[] splits = scaledDataSet.randomSplit(new double[]{0.7, 0.3});
JavaRDD trainingData = splits[0];
trainingData.cache();
trainingData.count(); 
...
// your other code

**Comment by ansarisam** [comment]
This is what I experimented with to see the effect:
1. Upgraded to 1.0.0-beta2
2. Added .rddTrainingApproach(RDDTrainingApproach.Direct)

Result: The saveAsText is no longer being called which is great. The network training process starts quickly. 
I notice that three processes are going in a loop:
- treeAggregate at ParameterAveragingTrainingMaster.java:663
- treeAggregate at ParameterAveragingTrainingMaster.java:663
- mapToPair at SparkUtils.java:415
- zipWithIndex at SparkUtils.java:415
- collect at SparkUtils.java:377
After 1500+ iteration it has not terminated yet.
Is there any way to know how many iteration of the above will go on?

I tried to keep a very simple network with the following configurations:
1. 3 layer
2. 5 nodes in each layer
3. AveragingFrequency: 100
4. Batchsize: 2000,000 
5. SaveUpdater(true)

Spark Cached the data with 28 partitions with a total size of 7GB
Total file size on disk is 9GB with 45million rows.

**Comment by AlexDBlack** [comment]
OK, so to reiterate what I've already said:

First: Yes, what you are seeing for the export is slow. For a rough comparison, I've processed about 400GB of text (not CSV) in about 30 minutes (22x 16 core cluster), about 10-20GB CSV on a 10 node cluster in about 20 minutes.

Second: Again, preprocess your data *once* - I told you how to do this here: https://github.com/deeplearning4j/deeplearning4j/issues/6437#issuecomment-421248606
That'll give you best performance and remove the need to preprocess your data on every single run.
Did you do this? If so, how long did it take to export?
If you're training with ```.rddTrainingApproach(RDDTrainingApproach.Direct)``` training *will* be slower than exporting first, unless all of your data fits in memory simultaneously (unlikely for 45m rows).

Third: as I already mentioned...
> Finally, you might consider using SharedTrainingMaster. Your problem isn't caused or related to ParameterAveragingTrainingMaster, but SharedTrainingMaster will give you better training performance once you get to that point.
https://github.com/deeplearning4j/dl4j-examples/blob/master/dl4j-spark-examples/dl4j-spark/src/main/java/org/deeplearning4j/patent/TrainPatentClassifier.java#L191-L197

See also: https://deeplearning4j.org/docs/latest/deeplearning4j-spark-parameter-server
I would recommend using the gradient sharing approach over parameter averaging. With a little configuration and appropriate networks, we see about linear speedup in practice vs. number of nodes. It should always perform better than parameter averaging due to significantly reduced communication requirements.


As for how long training will take with parameter averaging: It's reasonably straightforward to calculate the number of splits.
```numSplits = totalExamples / (avgFreq * minibatchSize * numExecutors)```
i.e., each of numExecutor executor does minibatchSize * avgFrequency examples before averaging.
That's the aggegate/map/collect loop you're seeing.

Finally, should you use averaging frequency of 100? No, that's probably too infrequent for proper network convergence, you'll want something smaller. Though it's a moot point if you are using SharedTrainingMaster/gradient sharing.

**Comment by AlexDBlack** [comment]
Closing this due to inactivity. We can reopen if needed.
